Exposing the emu_options (and slot options) to LUA

npwoods opened the pull request.

rb6502 commented:
Have you seen this in-progress PR which also exposes the slot options to Lua?

https://github.com/mamedev/mame/pull/7436

npwoods commented:
Actually I didn't see this at all, what a coincidence (!!!)

I'll need to review the other PR in detail.  Assuming it does the same thing (read/write access to slot options), I don't have a strong opinion.

npwoods commented:
Reviewing the other PR, I have two observations:
* PR#7436 seems to only provide read only access to the slot options, whereas PR#7458 provides the ability to change them
* PR#7458 tries to keep the LUA objects closer to the internal MAME object model.

Ultimately, I don't really care which approach is taken, but if PR#7436 gets committed I will need to follow up with extensions to provide the ability to change the slot values.

npwoods commented:
Any thoughts on #7436 vs #7458?

npwoods commented:
I've updated this PR to work with the latest MAME, even though LUA support seems to be hosed (the console plugin doesn't work)

npwoods commented:
...bueller???

cuavas commented:
I’m doing some fairly major refactoring on the Lua engine code following on from upgrading to sol3 with the move to C++17.  Although what I’m doing isn’t directly related to slots, it *will* conflict with this stuff.  I need a bit more time to get some more stuff in before it’s ready to push, then we can re-asses this.

npwoods commented:
Thanks Vas!

cuavas commented:
As expected, this now conflicts with mainline.  In particular, the proper way to expose what was a `device_interface_iterator` but is now called a `device_interface_enumerator` has changed.

npwoods commented:
Updates

cuavas commented on src/frontend/mame/luaengine.cpp:
```diff
@@ -987,6 +987,38 @@ void lua_engine::initialize()
+ * options.entries[] - get table of option entries (k=name, v=core_options::entry)
+ */
+
+	auto emu_options_type = sol().registry().new_usertype<emu_options>("new", sol::no_constructor, sol::base_classes, sol::bases<core_options>());
+	emu_options_type.set("slot_option", [](emu_options &opts, const std::string &name) { return opts.find_slot_option(name); });
+	sol().registry().set("emu_options", emu_options_type);
```
Please use the more modern syntax for additions.  You can see how I’ve been changing stuff as I go.

Also, please keep the two blank lines between each user type.

cuavas commented on src/frontend/mame/luaengine.cpp:
```diff
@@ -987,6 +987,38 @@ void lua_engine::initialize()
+ * manager:options():slot_option("name")
+ * manager:machine():options():slot_option("name")
+ *
+ * slot_option:value()
+ * slot_option:specified_value()
+ * slot_option:bios()
+ * slot_option:default_card_software()
```
At least some of these would probably work better as read-only properties.  Also, try to give a description for at least some of them.

cuavas commented on src/frontend/mame/luaengine.cpp:
```diff
@@ -987,6 +987,38 @@ void lua_engine::initialize()
+	slot_option_type.set("specified_value", &slot_option::specified_value);
+	slot_option_type.set("bios", &slot_option::bios);
+	slot_option_type.set("default_card_software", &slot_option::default_card_software);
+	slot_option_type.set("specify", [](slot_option &opt, std::string &&text) { opt.specify(std::move(text)); });
```
You haven’t documented this at all.

cuavas replied:
Should this be a method, or should the `value` property be writable?  Maybe it needs to be a method because of the BIOS option.  But speaking of the BIOS option, maybe it should take two arguments, and if the second argument is not supplied or is Lua `nil` not put the `,bios=` part in.  We don’t really need to expose that particular piece of ugliness to Lua.

Also, for all the class wrappers we’ve been putting methods before properties, so if this remains a method, it should come before the properties for consistency.

cuavas commented on src/frontend/mame/luaengine.cpp:
```diff
@@ -1786,21 +1819,53 @@ void lua_engine::initialize()
+/*  device_slot_interface library
+ *
+ * manager:machine().slots[slot_name]
+ *
+ * slot:fixed()
+ * slot:has_selectable_options()
+ *
+ * slot.options[] - get options table (k=name, v=device_slot_interface::slot_option)
+ */
+
+	auto slot_type = sol().registry().new_usertype<device_slot_interface>("new", sol::no_constructor);
+	slot_type.set("fixed", &device_slot_interface::fixed);
+	slot_type.set("has_selectable_options", &device_slot_interface::has_selectable_options);
+	slot_type.set("default_option", &device_slot_interface::default_option);
+	slot_type.set("options", sol::property([this](device_slot_interface &slot) {
+			sol::table option_table = sol().create_table();
+			for (auto &x : slot.option_list())
+				option_table[x.first] = &*x.second;
+			return option_table;
+		}));
+	sol().registry().set("slot", slot_type);
```
Once again, documentation, syntax, spacing, methods vs properties.  Also, it’s better to make a lightweight wrapper for the options rather than walking it and building a Lua table every time.

npwoods commented:
I believe that I've addressed all of your feedback other than this:

_> Also, it’s better to make a lightweight wrapper for the options rather than walking it and building a Lua table every time._

Can you point to something in luaengine.cpp that does this well, as a model to follow?

cuavas commented:
> 
> 
> I believe that I've addressed all of your feedback other than this:
> 
> _> Also, it’s better to make a lightweight wrapper for the options rather than walking it and building a Lua table every time._
> 
> Can you point to something in luaengine.cpp that does this well, as a model to follow?

In luaengine.cpp, luaengine_mem.cpp and luaengine_render.ipp there are Lua container wrappers for various things.  You basically specialise `sol::is_container` and `sol::usertype_container`.  Most of them allow lookups by key using `[k]` and by index using `:at(i)`.

cuavas commented on src/frontend/mame/luaengine.cpp:
```diff
@@ -1786,21 +1820,55 @@ void lua_engine::initialize()
+ * slot.options[] - get options table (k=name, v=device_slot_interface::slot_option)
+ */
+
+	auto slot_type = sol().registry().new_usertype<device_slot_interface>("new", sol::no_constructor);
+	slot_type["fixed"] = sol::property(&device_slot_interface::fixed);
+	slot_type["has_selectable_options"] = sol::property(&device_slot_interface::has_selectable_options);
+	slot_type.set("default_option", &device_slot_interface::default_option);
+	slot_type.set("options", sol::property([this](device_slot_interface &slot) {
+			sol::table option_table = sol().create_table();
+			for (auto &x : slot.option_list())
+				option_table[x.first] = &*x.second;
+			return option_table;
+		}));
+	sol().registry().set("slot", slot_type);
```
You’re still using the old syntax for this – you can supply the name when creating the user type, you don’t need the `"new"` for constructorless classes.  You’re mixing `set` and `[]` syntax in a single class (same in other places).

cuavas commented on src/frontend/mame/luaengine.cpp:
```diff
@@ -1786,21 +1820,55 @@ void lua_engine::initialize()
+ * slot_option:selectable()
+ */
+
+	auto dislot_option_type = sol().registry().new_usertype<device_slot_interface::slot_option>("new", sol::no_constructor);
+	dislot_option_type.set("selectable", &device_slot_interface::slot_option::selectable);
+	sol().registry().set("dislot_option", dislot_option_type);
```
This is also using old syntax.

If you’re exposing this thing, why not also expose the shortname and description of the slot option’s device type, so the script can generate a list like `-listslots`?  You can expose the default BIOS and clock frequency (remember this can contain magic numbers for ratio of slot clock), too, for informational purposes.

cuavas commented on src/frontend/mame/luaengine.cpp:
```diff
@@ -1778,6 +1814,37 @@ void lua_engine::initialize()
+	slot_type["fixed"] = sol::property(&device_slot_interface::fixed);
+	slot_type["has_selectable_options"] = sol::property(&device_slot_interface::has_selectable_options);
+	slot_type["default_option"] = sol::property(&device_slot_interface::default_option);
+	slot_type["options"] = sol::property([](const device_slot_interface &slot) { return lua_engine::standard_tag_object_ptr_map<device_slot_interface::slot_option>(slot.option_list()); });
```
You don’t need `lua_engine::` from inside the `lua_engine` class.  Same applies in other places.

cuavas commented on src/frontend/mame/luaengine.cpp:
```diff
@@ -1778,6 +1814,37 @@ void lua_engine::initialize()
+
+	auto dislot_option_type = sol().registry().new_usertype<device_slot_interface::slot_option>("dislot_option", "new", sol::no_constructor);
+	dislot_option_type["selectable"] = sol::property(&device_slot_interface::slot_option::selectable);
+	dislot_option_type["default_bios"] = sol::property([](const device_slot_interface::slot_option &opt) { return opt.default_bios(); });
+	dislot_option_type["clock"] = sol::property([](const device_slot_interface::slot_option &opt) { return opt.clock(); });
```
You can do this with a static cast rather than making a lambda wrapper – see e.g. `device_type["subdevice"]` and device_type["siblingdevice"]` which use static casts to choose between overloaded methods.

cuavas commented on src/frontend/mame/luaengine_mem.cpp:
```diff
@@ -652,9 +538,9 @@ void lua_engine::initialize_memory()
-	memory_type["banks"] = sol::property([] (memory_manager &mm) { return standard_memory_object_map<memory_bank>(mm.banks()); });
-	memory_type["regions"] = sol::property([] (memory_manager &mm) { return standard_memory_object_map<memory_region>(mm.regions()); });
-	memory_type["shares"] = sol::property([] (memory_manager &mm) { return standard_memory_object_map<memory_share>(mm.shares()); });
+	memory_type["banks"] = sol::property([] (memory_manager &mm) { return lua_engine::standard_tag_object_ptr_map<memory_bank>(mm.banks()); });
+	memory_type["regions"] = sol::property([] (memory_manager &mm) { return lua_engine::standard_tag_object_ptr_map<memory_region>(mm.regions()); });
+	memory_type["shares"] = sol::property([] (memory_manager &mm) { return lua_engine::standard_tag_object_ptr_map<memory_share>(mm.shares()); });
```
These shouldn’t need `lua_engine::` since `standard_tag_object_ptr_map` is not an argument-dependent name.
